ipa-core/src/protocol/ipa_prf/malicious_security/lagrange.rs

```diff
@@ -0,0 +1,240 @@
+/// The `LagrangeTable` also specifies `M` "x coordinates" for the output points
+/// The "x coordinates" of the output points are `x_N` to `x_(N+M-1)` are `N*F::ONE` to `(N+M-1)*F::ONE`.
+pub struct LagrangeTable<F: Field, N: ArrayLength, M: ArrayLength> {
+    table: GenericArray<GenericArray<F, N>, M>,
+}
+
+impl<F, N> LagrangeTable<F, N, U1>
+where
+    F: Field,
+    N: ArrayLength,
+{
+    /// generates a `CanonicalLagrangeTable` from `CanoncialLagrangeDenominators` for a single output point
+    /// The "x coordinate" of the output point is `x_output`.
+    pub fn new(denominator: &CanonicalLagrangeDenominator<F, N>, x_output: &F) -> Self {
+        let mut table = denominator.denominator.clone();
+        compute_table_row(x_output, &mut table);
+        LagrangeTable::<F, N, U1> {
+            table: GenericArray::from_array([table; 1]),
+        }
+    }
+}
+
+impl<F, N, M> LagrangeTable<F, N, M>
+where
+    F: Field,
+    N: ArrayLength,
+    M: ArrayLength,
+{
+    /// This function uses the `LagrangeTable` to evaluate `polynomial` on the specified output "x coordinates"
+    /// outputs the "y coordinates" such that `(x,y)` lies on `polynomial`
+    pub fn eval(&self, polynomial: &Polynomial<F, N>) -> GenericArray<F, M> {
+        let mut result = iter::repeat(F::ONE)
+            .take(M::USIZE)
+            .collect::<GenericArray<_, _>>();
+        self.mult_result_by_evaluation(polynomial, &mut result);
+        result
+    }
+
+    /// This function uses the `LagrangeTable` to evaluate `polynomial` on the specified output "x coordinates"
+    /// the "y coordinates" of the evaluation are multiplied to `result`
+    pub fn mult_result_by_evaluation(
+        &self,
+        polynomial: &Polynomial<F, N>,
+        result: &mut GenericArray<F, M>,
+    ) {
+        for (y, base) in result.iter_mut().zip(self.table.iter()) {
+            *y *= base
+                .iter()
+                .zip(polynomial.y_coordinates.iter())
+                .fold(F::ZERO, |acc, (&base, &y)| acc + base * y);
+        }
+    }
+}
+
+impl<F, N, M> From<CanonicalLagrangeDenominator<F, N>> for LagrangeTable<F, N, M>
+where
+    F: Field,
+    N: ArrayLength,
+    M: ArrayLength,
+{
+    fn from(value: CanonicalLagrangeDenominator<F, N>) -> Self {
+        // assertion that field is large enough
+        // also checks that `try_from` for conversions from sufficiently small `u128` to `F` do not panic
+        debug_assert!(F::BITS > usize::BITS - N::USIZE.leading_zeros() - M::USIZE.leading_zeros());
+
+        let mut table = iter::repeat(value.denominator.clone())
+            .take(M::USIZE)
+            .collect::<GenericArray<_, _>>();
+        table.iter_mut().enumerate().for_each(|(i, row)| {
+            compute_table_row(&F::try_from((i + N::USIZE) as u128).unwrap(), row);
+        });
+        LagrangeTable { table }
+    }
+}
+
+/// helper function to compute a single row of `CanonicalLagrangeTable`
```

Comment:
```suggestion
/// helper function to compute a single row of [`LagrangeTable`]
/// [`LagrangeTable`]: LagrangeTable
```

in this case you won't miss updating comments after renaming